[ALLUXIO-3178] Worker send configuration when register with master

LuQQiu opened the pull request:
https://alluxio.atlassian.net/projects/ALLUXIO/issues/ALLUXIO-3178

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19410/<h2>Build result: FAILURE</span></h2>[...truncated 2102 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/target/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-keyvalue-common/1.8.0-SNAPSHOT/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/checker/pom.xml to org.alluxio/alluxio-checker/1.8.0-SNAPSHOT/alluxio-checker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/pom.xml to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/pom.xml to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/worker/pom.xml to org.alluxio/alluxio-core-server-worker/1.8.0-SNAPSHOT/alluxio-core-server-worker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/logserver/pom.xml to org.alluxio/alluxio-logserver/1.8.0-SNAPSHOT/alluxio-logserver-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/pom.xml to org.alluxio/alluxio-core-client/1.8.0-SNAPSHOT/alluxio-core-client-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/pom.xml to org.alluxio/alluxio-core/1.8.0-SNAPSHOT/alluxio-core-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/proxy/pom.xml to org.alluxio/alluxio-core-server-proxy/1.8.0-SNAPSHOT/alluxio-core-server-proxy-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/tests/pom.xml to org.alluxio/alluxio-tests/1.8.0-SNAPSHOT/alluxio-tests-1.8.0-SNAPSHOT.pomchannel stoppedArchiving artifacts
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19413/<h2>Build result: FAILURE</span></h2>[...truncated 2391 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/pom.xml to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/pom.xml to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/worker/pom.xml to org.alluxio/alluxio-core-server-worker/1.8.0-SNAPSHOT/alluxio-core-server-worker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/logserver/pom.xml to org.alluxio/alluxio-logserver/1.8.0-SNAPSHOT/alluxio-logserver-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/pom.xml to org.alluxio/alluxio-core-client/1.8.0-SNAPSHOT/alluxio-core-client-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/pom.xml to org.alluxio/alluxio-core/1.8.0-SNAPSHOT/alluxio-core-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/proxy/pom.xml to org.alluxio/alluxio-core-server-proxy/1.8.0-SNAPSHOT/alluxio-core-server-proxy-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/proxy/target/alluxio-core-server-proxy-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-server-proxy/1.8.0-SNAPSHOT/alluxio-core-server-proxy-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/proxy/target/alluxio-core-server-proxy-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-server-proxy/1.8.0-SNAPSHOT/alluxio-core-server-proxy-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/tests/pom.xml to org.alluxio/alluxio-tests/1.8.0-SNAPSHOT/alluxio-tests-1.8.0-SNAPSHOT.pomchannel stoppedArchiving artifacts
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19414/
Test PASSed.

LuQQiu commented:
@aaudiber @apc999 @gpang  PTAL, thanks!
I mainly add worker-related configurations to worker register and set up the main structure of ServerConfigurationReport.java.

gpang commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -73,6 +77,21 @@
+  /** Prefixes that are not related to workers but exists in worker-related scope.*/
+  private static final String[] WORKER_NOT_RELATED_PREFIX = {"alluxio.fuse", "alluxio.proxy",
+      "alluxio.integration.master", "alluxio.integration.mesos.master"};
+  /** Properties that are not related to workers but exist in worker-related scope.*/
```
If these are not supposed to be in the worker's scope, then shouldn't we just remove them from the worker scope?

LuQQiu replied:
Those properties have the range Scope.ALL.  For Scope=worker(Property is used in workers), Scope=SERVER(Property is used in masters and workers) we are clear that those properties are related to workers. 

But for Scope=ALL which is the default scope, many properties are integration related properties that may/may not related to workers.

aaudiber replied:
If a property does not apply to workers and its scope is set to `Scope.ALL`, the scope is set incorrectly and should be fixed. Mesos integration properties are not used by clients or servers, so they should have `Scope.NONE`. The master properties only apply to master, so they should have scope `Scope.MASTER`

gpang commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -604,5 +623,44 @@ public static void validate() {
+   * @param keyName the name of the PropertyKey to check
+   * @return if the PropertyKey is worker related
+   */
+  private static boolean isWorkerRelated(String keyName) {
```
Why can't we just use the worker scope for this type of check?

gpang replied:
I think if you use `Scope#contains()`, that should be what you want. Ideally, just checking for `Scope#contains(WORKER)` should be all that we need to do.

gpang commented on core/server/master/src/main/java/alluxio/master/block/BlockMasterWorkerServiceHandler.java:
```diff
@@ -134,6 +136,13 @@ public RegisterWorkerTResponse registerWorker(final long workerId,
       public RegisterWorkerTResponse call() throws AlluxioException {
         mBlockMaster.workerRegister(workerId, storageTiers, totalBytesOnTiers, usedBytesOnTiers,
             currentBlocksOnTiers);
+        if (options.isSetConfigList()) {
```
I think this processing should be in `BlockMaster`, instead of this rpc handler class.

gpang commented on core/server/master/src/main/java/alluxio/master/meta/ServerConfigurationReport.java:
```diff
@@ -0,0 +1,103 @@
+ * This class is responsible for containing configurations and creating server configuration report.
+ */
+public final class ServerConfigurationReport {
+  private static Map<Long, List<ConfigProperty>> sWorkerConfMap;
```
Do we have to use static data structures for this?

LuQQiu replied:
We can create a ServerConfigurationReport instance when a master gets leadership. I will change it.

aaudiber replied:
the configuration checking should be owned by either the root master or meta master.

LuQQiu replied:
The server configuration checker is owned by AlluxioMasterProcess now. Thanks!

aaudiber commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -604,5 +623,44 @@ public static void validate() {
+        String sourceStr;
+        if (source == Source.SITE_PROPERTY) {
+          sourceStr =
+              String.format("%s (%s)", source.name(), getSitePropertiesFile());
```
we have similar logic in other places for converting a source to a sourceStr. Can we consolidate them?

LuQQiu replied:
Create method getFormattedSource(PropertyKey key) to consolidate duplicate codes. thanks!

aaudiber commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -604,5 +623,44 @@ public static void validate() {
+  /**
+   * @return a list of worker-related configurations
+   */
+  public static List<ConfigProperty> getWorkerConfiguration() {
```
could we generate this by iterating over all property keys and taking the ones with scope worker? java 8 stream API should let us do this in a single statement

LuQQiu replied:
We do not want to check Scope = Server? like alluxio.underfs.address?

aaudiber replied:
`key.getScope().contains(Scope.WORKER)`

aaudiber commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -73,6 +77,21 @@
+  /** Use to filter property keys that start with fs.*/
+  private static final String FS_CONF_PREFIX = "fs";
+  /** Prefixes that are not related to workers but exists in worker-related scope.*/
+  private static final String[] WORKER_NOT_RELATED_PREFIX = {"alluxio.fuse", "alluxio.proxy",
```
fuse is Scope.NONE, proxy is Scope.NONE, integration is scope.NONE

LuQQiu replied:
Many of the Fuse, Proxy, and Integration properties have the ConsistencyCheckLevel = WARN and even ENFORCE(like "alluxio.integration.mesos.alluxio.jar.url", ""alluxio.integration.mesos.master.name")

If we set them as Scope.NONE, then we will not check those properties at all(both in masters and workers).

aaudiber replied:
Good call, for the properties I mentioned that do not apply to Clients, Workers, or Masters, we might as well have `ConsistencyCheckLevel.IGNORE`

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/ServerConfigurationReport.java:
```diff
@@ -0,0 +1,103 @@
+/**
+ * This class is responsible for containing configurations and creating server configuration report.
+ */
+public final class ServerConfigurationReport {
```
perhaps rename to `ServerConfigurationChecker`? Instances of this class aren't themselves reports

LuQQiu replied:
Much better name, changed, thanks

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/ServerConfigurationReport.java:
```diff
@@ -0,0 +1,103 @@
+/**
+ * This class is responsible for containing configurations and creating server configuration report.
+ */
+public final class ServerConfigurationReport {
```
can we write this class in a general way so that it can be applied to any type of scope? Then we don't need to check isMaster and duplicate logic

LuQQiu replied:
I create a ConfigRecorder to store lockers, confMap, lostConfMap. 
and init one MasterConfigRecorder and one WorkerConfigRecorder in the ServerConfigurationChecker class. 

I am not quite sure is that the general way? or are there better ways to process the duplicate logic? @aaudiber

aaudiber commented on core/server/worker/src/main/java/alluxio/worker/block/BlockMasterClient.java:
```diff
@@ -137,18 +139,20 @@ public Command call() throws TException {
+    retryRPC(() -> {
+      RegisterWorkerTOptions options = new RegisterWorkerTOptions();
+      options.setConfigList(configList.stream()
+          .map(ConfigProperty :: toThrift).collect(Collectors.toList()));
```
google style is no spaces around the `::` in a method reference
https://google.github.io/styleguide/javaguide.html#s4.6.2-horizontal-whitespace

LuQQiu replied:
Changed, thanks for the info!

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19435/<h2>Build result: FAILURE</span></h2>[...truncated 2102 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/target/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-keyvalue-common/1.8.0-SNAPSHOT/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/checker/pom.xml to org.alluxio/alluxio-checker/1.8.0-SNAPSHOT/alluxio-checker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/pom.xml to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/pom.xml to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/worker/pom.xml to org.alluxio/alluxio-core-server-worker/1.8.0-SNAPSHOT/alluxio-core-server-worker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/logserver/pom.xml to org.alluxio/alluxio-logserver/1.8.0-SNAPSHOT/alluxio-logserver-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/pom.xml to org.alluxio/alluxio-core-client/1.8.0-SNAPSHOT/alluxio-core-client-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/pom.xml to org.alluxio/alluxio-core/1.8.0-SNAPSHOT/alluxio-core-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/proxy/pom.xml to org.alluxio/alluxio-core-server-proxy/1.8.0-SNAPSHOT/alluxio-core-server-proxy-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/tests/pom.xml to org.alluxio/alluxio-tests/1.8.0-SNAPSHOT/alluxio-tests-1.8.0-SNAPSHOT.pomchannel stoppedArchiving artifacts
Test FAILed.

LuQQiu commented on core/server/master/src/main/java/alluxio/master/meta/ServerConfigurationReport.java:
```diff
@@ -0,0 +1,103 @@
+    if (isMaster) {
+      configList = sLostMasterConfMap.get(id);
+      if (configList != null) {
+        sMasterConfMap.put(id, configList);
```
I have a ```private ConcurrentHashMap<Long, Object> mLockers;``` to save the lockers, each time when we registerNewConf or addConf/removeConf, we get the locker from this map by using:

```
      mLockers.putIfAbsent(id, new Object());
      synchronized (mLockers.get(id)) {
```

Is it the right way to make those methods threadsafe and concurrent? @gpang

LuQQiu commented:
@aaudiber @gpang I have addressed the comments and please take a look, thanks!

The main changes are

1. init Server Configuration Checker in AlluxioMasterProcess and add callbacks to enable DefaultBlockMaster to registerNewConf/addConf/removeConf to mConfigServer inside AlluxioMasterProcess.

2. Make server configuration checker more general by introducing ConfigRecorder and add lockers to methods inside ConfigRecorder.

3. instead of having a isWorkerRelated() method when getWorkerConfiguration(), change the unrelated property key scope to NONE or MASTER and use key.getScope.contains(Scope.WORKER).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19436/
Test PASSed.

aaudiber commented on core/server/master/src/main/java/alluxio/master/block/BlockMaster.java:
```diff
@@ -190,4 +196,25 @@ Command workerHeartbeat(long workerId, Map<String, Long> usedBytesOnTiers,
+   *
+   * @param function the function to save to corresponding callback list
+   */
+  void workerAddConfCallback(Consumer<Long> function);
```
A common naming pattern is "register{ACTION}Listener", where {ACTION} is the action that will trigger the callback. For example, `registerLostWorkerFoundListener`

LuQQiu replied:
Changed to : 
```
registerLostWorkerFoundListener
 registerWorkerLostListener
registerNewWorkerConfListener
```
Thanks for the callback conventions.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19438/
Test PASSed.

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/conf/ConfigRecorder.java:
```diff
@@ -0,0 +1,86 @@
+ */
+public class ConfigRecorder {
+  /** Map from ids to lockers. */
+  private ConcurrentHashMap<Long, Object> mLockers;
```
Could we just make every method synchronized? Configuration change is a rare event, and all of these methods are fast

LuQQiu replied:
Change to
```
private Object mLocker;
synchronized (mLocker)  in all methods
```

Use one object to synchronize all the methods inside a ConfigRecorder.

aaudiber replied:
You don't need a separate lock to do this, you can just use `synchronized (this) {`,  or even simpler, add the `synchronized` keyword to the method signatures. These are all the same:
```java
  private Object mLock;
  public void method() {
    synchronized (mLock) {
      doSomething();
    }
  }
```
```java
  public void method() {
    synchronized (this) {
      doSomething();
    }
  }
```
```java
  public synchronized void method() {
    doSomething();
  }
```

aaudiber commented on core/common/src/main/java/alluxio/Configuration.java:
```diff
@@ -604,5 +623,22 @@ public static void validate() {
+   * @return a list of worker-related configurations
+   */
+  public static List<ConfigProperty> getWorkerConfiguration() {
+    List<ConfigProperty> configInfoList = new ArrayList<>();
```
Can we write a method that returns the set of all property keys? Then this would be 
```java
return getAllPropertyKeys().stream()
  .filter(key -> key.getScope().contains(Scope.WORKER))
  .map(key -> new ConfigProperty().setName(key.getName()).setValue(get(key))
      .setSource(getFormattedSource(key))).collect()
```

LuQQiu replied:
Changed to : 
```
  public static List<ConfigProperty> getWorkerConfiguration() {
    return toMap().keySet().stream().map(PropertyKey::fromString)
        .filter(key -> key.getScope().contains(Scope.WORKER))
        .map(key -> new ConfigProperty().setName(key.getName())
            .setValue(get(key)).setSource(getFormattedSource(key)))
        .collect(Collectors.toList());
  }
```

Add one ```.map(PropertyKey::fromString) ```so that we do not need to write a separate method.

LuQQiu commented on core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:
```diff
@@ -162,6 +166,16 @@ public Object getFieldValue(MasterWorkerInfo o) {
       new IndexedSet<>(ID_INDEX, ADDRESS_INDEX);
 
+  /** Saves the add configuration call back functions. */
+  private final List<Consumer<Long>> mAddConfCallbacks = new ArrayList<>();
```
We only need to write once and read many times. I thought of the need of using a concurrent LinkedBlockingQueue, but many functions of LinkedBlockingQueue is wasteful.

LuQQiu commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -111,6 +116,9 @@
   protected final SafeModeManager mSafeModeManager;
 
+  /** The server-side configuration checker. */
+  private ServerConfigurationChecker mConfigChecker;
```
Done, thanks!

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19440/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19441/<h2>Build result: FAILURE</span></h2>[...truncated 2101 lines...][JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/keyvalue/common/target/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-keyvalue-common/1.8.0-SNAPSHOT/alluxio-keyvalue-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/checker/pom.xml to org.alluxio/alluxio-checker/1.8.0-SNAPSHOT/alluxio-checker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/pom.xml to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/fs/target/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-client-fs/1.8.0-SNAPSHOT/alluxio-core-client-fs-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/pom.xml to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/common/target/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar to org.alluxio/alluxio-core-common/1.8.0-SNAPSHOT/alluxio-core-common-1.8.0-SNAPSHOT-tests.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/worker/pom.xml to org.alluxio/alluxio-core-server-worker/1.8.0-SNAPSHOT/alluxio-core-server-worker-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/logserver/pom.xml to org.alluxio/alluxio-logserver/1.8.0-SNAPSHOT/alluxio-logserver-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/client/pom.xml to org.alluxio/alluxio-core-client/1.8.0-SNAPSHOT/alluxio-core-client-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/pom.xml to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/examples/target/alluxio-examples-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-examples/1.8.0-SNAPSHOT/alluxio-examples-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/pom.xml to org.alluxio/alluxio-core/1.8.0-SNAPSHOT/alluxio-core-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/core/server/proxy/pom.xml to org.alluxio/alluxio-core-server-proxy/1.8.0-SNAPSHOT/alluxio-core-server-proxy-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/pom.xml to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.pom[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/shell/target/alluxio-shell-1.8.0-SNAPSHOT-sources.jar to org.alluxio/alluxio-shell/1.8.0-SNAPSHOT/alluxio-shell-1.8.0-SNAPSHOT-sources.jar[JENKINS] Archiving /home/jenkins/workspace/Alluxio-Pull-Request-Builder/tests/pom.xml to org.alluxio/alluxio-tests/1.8.0-SNAPSHOT/alluxio-tests-1.8.0-SNAPSHOT.pomchannel stoppedArchiving artifacts
Test FAILed.

LuQQiu commented:
@gpang @aaudiber PLTAL, thanks!
I modified the synchronized methods and add a configRecorder test.
For serverConfigurationChecker and worker register with integration test, it may be better to add them after finishing generateReport() method of serverConfigurationChecker.  I plan to leave it to other PR.

aaudiber commented on core/server/master/src/test/java/alluxio/master/meta/ConfigRecorderTest.java:
```diff
@@ -0,0 +1,99 @@
+
+    Map<Long, List<ConfigProperty>> confMap = configRecorder.getConfMap();
+
+    Assert.assertTrue(confMap.containsKey(mIdOne));
```
can shorten the asserts with static imports

aaudiber commented on core/server/master/src/test/java/alluxio/master/meta/ConfigRecorderTest.java:
```diff
@@ -0,0 +1,99 @@
+  public void detectNodeLost() {
+    ConfigRecorder configRecorder = createConfigRecorder();
+
+    configRecorder.detectNodeLost(mIdOne);
```
the name `detectNodeLost` sounds like a method that would check whether a node has been lost. Perhaps rename to `handleNodeLost` or `nodeLost`?

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/ConfigRecorder.java:
```diff
@@ -0,0 +1,82 @@
+   * @return the configuration map of live masters or workers
+   */
+  public synchronized Map<Long, List<ConfigProperty>> getConfMap() {
+    return mConfMap;
```
would it make sense to return an immutable view here? Otherwise things could get confusing if the caller modifies the returned map.

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/ConfigRecorder.java:
```diff
@@ -0,0 +1,82 @@
+  }
+
+  /**
+   * Removes the configuration of a master/worker from ConfMap to LostConfMap
```
ConfMap and LostConfMap are implementation details that don't need to be exposed to users of the class, they shouldn't be part of the javadoc

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/ConfigRecorder.java:
```diff
@@ -0,0 +1,82 @@
+ * This class is responsible for containing server-side configurations.
+ */
+public class ConfigRecorder {
+  /** Stores the configuration of a master/worker with Long id. */
```
Nothing in the implementation is specific to masters, workers, or servers. We can just say "node" to keep it general-purpose

aaudiber commented on core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:
```diff
@@ -672,6 +688,9 @@ public long getWorkerId(WorkerNetAddress workerNetAddress) {
         mWorkers.add(lostWorker);
         mLostWorkers.remove(lostWorker);
+        for (Consumer<Long> function : mAddConfCallbacks) {
+          function.accept(lostWorker.getId());
```
how will we handle the case where a worker is restarted with new configuration?

LuQQiu replied:
If a worker gets its old id and restarts with new configuration, it will register with new configuration. When this worker register, we will overwrite its previous configuration in the confMap and remove its old configuration in lostConfMap if exists.

If the worker gets a new id and register with new configuration, the previous old id worker will be detected as lost worker and move to LostConfMap. new worker id alone with the worker configuration will be add to ConfMap.

aaudiber commented on core/server/master/src/main/java/alluxio/master/block/BlockMaster.java:
```diff
@@ -190,4 +196,25 @@ Command workerHeartbeat(long workerId, Map<String, Long> usedBytesOnTiers,
   void reportLostBlocks(List<Long> blockIds);
+
+  /**
+   * Saves the callback function to use when lost workers become alive.
```
Use the same verb as the method if they are synonyms, e.g. "Registers a callback function ..."

aaudiber commented on core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:
```diff
@@ -162,6 +168,16 @@ public Object getFieldValue(MasterWorkerInfo o) {
       new IndexedSet<>(ID_INDEX, ADDRESS_INDEX);
 
+  /** Saves the add configuration call back functions. */
+  private final BlockingQueue<Consumer<Long>> mAddConfCallbacks = new LinkedBlockingQueue<>();
```
mLostWorkerFoundListeners

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19442/
Test PASSed.

aaudiber commented on core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:
```diff
@@ -162,6 +168,16 @@ public Object getFieldValue(MasterWorkerInfo o) {
+  private final BlockingQueue<Consumer<Long>> mAddConfCallbacks = new LinkedBlockingQueue<>();
+
+  /** Saves the remove configuration call back functions. */
+  private final BlockingQueue<Consumer<Long>> mRemoveConfCallbacks = new LinkedBlockingQueue<>();
```
mWorkerLostListeners

aaudiber commented on core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:
```diff
@@ -162,6 +168,16 @@ public Object getFieldValue(MasterWorkerInfo o) {
+  private final BlockingQueue<Consumer<Long>> mRemoveConfCallbacks = new LinkedBlockingQueue<>();
+
+  /** Saves the register new configuration call back functions. */
+  private final BlockingQueue<BiConsumer<Long, List<ConfigProperty>>> mRegisterConfCallbacks
```
`mWorkerRegisteredListeners`

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/ServerConfigurationChecker.java:
```diff
@@ -0,0 +1,75 @@
+ */
+public final class ServerConfigurationChecker {
+  private ConfigRecorder mMasterConfigRecorder;
+  private ConfigRecorder mWorkerConfigRecorder;
```
instead of storing two `ConfigRecorders` in this class and adding an `isMaster` parameter to every method, could we instead have two instances of `ServerConfigurationChecker`, one that we use for masters and one that we use for workers? Then `ServerConfigurationChecker` would only hold one `ConfigRecorder` and we wouldn't need any of the `isMaster` stuff.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19445/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19446/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19447/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19449/
Test PASSed.

LuQQiu commented:
@aaudiber @gpang I changed ServerConfigurationChecker to be more general and change names. PTAL, thanks!

LuQQiu commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -111,6 +116,9 @@
   protected final SafeModeManager mSafeModeManager;
 
+  /** The worker configuration checker. */
+  private final ServerConfigurationChecker mWorkerConfigChecker;
```
I think it would be better to init configuration in startMasters and in initializing AlluxioMasterProcess.  Are there a need to init it again in stopMasters?

I am not quite sure where should I init configuration so that it can be init when a master become primary.
```
  public ServerConfigurationChecker() {
    init();
  }

  public void init() {
    mConfMap = new HashMap<>();
    mLostConfMap = new HashMap<>();
  }
```
Thanks!

gpang replied:
If there is a `reset` or `init` for the `ServerConfigurationChecker`, then in `stopMasters` you can just reset the state there.

LuQQiu replied:
Add a init function in ServerConfigurationChecker
```
  /**
   * Initializes the default {@link ServerConfigurationChecker}.
   */
  public synchronized void init() {
    mConfMap = new HashMap<>();
    mLostNodes = new HashSet<>();
  }
```

Add call this function in AlluxioMasterProcess.stopMasters()
```
  protected void stopMasters() {
    try {
      mWorkerConfigChecker.init();
      mRegistry.stop();
    } catch (IOException e) {
      throw Throwables.propagate(e);
    }
  }
```
Thanks!

aaudiber replied:
Actually I agree with Lu that it's better to call `init` during start instead of during stop. It's better to initialize closer to when we need it so that there's less time for something to interfere. When we call stop() there could still be outstanding RPC threads trying to register workers and modifying the server config checker

LuQQiu commented on core/server/master/src/main/java/alluxio/master/meta/ServerConfigurationChecker.java:
```diff
@@ -0,0 +1,81 @@
+ */
+public class ServerConfigurationChecker {
+  /** Map from a node id to its configuration. */
+  private Map<Long, List<ConfigProperty>> mConfMap;
```
So we only maintain one ConfMap and one lostIdSet to make everything simpler?

gpang commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -90,6 +92,9 @@
   private final MasterRegistry mRegistry;
 
+  /** The block master. */
+  private final BlockMaster mBlockMaster;
```
I don't think we need to save this. We can just use the master registry to get the block master.

LuQQiu commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -421,4 +431,31 @@ protected void stopServing() throws Exception {
+  /**
+   * Updates the config checker when a lost worker becomes alive.
+   *
+   * @param id the id of the worker to use
```
Done, thanks

LuQQiu commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -421,4 +431,31 @@ protected void stopServing() throws Exception {
+  /**
+   * Updates the config checker when a live worker becomes lost.
+   *
+   * @param id the id of the worker to use
```
Done

LuQQiu commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -421,4 +431,31 @@ protected void stopServing() throws Exception {
+  /**
+   * Updates the config checker when a worker registers with configuration.
+   *
+   * @param id the id of the worker to use
```
Done, thanks

LuQQiu commented on core/server/master/src/main/java/alluxio/master/block/BlockMaster.java:
```diff
@@ -190,4 +196,25 @@ Command workerHeartbeat(long workerId, Map<String, Long> usedBytesOnTiers,
+  /**
+   * Registers a callback function to use when lost workers become alive.
+   *
+   * @param function the function to save to corresponding callback list
```
I will take care of avoiding implementation details in the method docs. Thanks!

LuQQiu commented on core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:
```diff
@@ -162,6 +168,17 @@ public Object getFieldValue(MasterWorkerInfo o) {
   private final IndexedSet<MasterWorkerInfo> mLostWorkers =
       new IndexedSet<>(ID_INDEX, ADDRESS_INDEX);
 
+  /** Saves the add configuration call back functions. */
```
Done

LuQQiu commented on core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:
```diff
@@ -162,6 +168,17 @@ public Object getFieldValue(MasterWorkerInfo o) {
+  private final BlockingQueue<Consumer<Long>> mLostWorkerFoundListeners
+      = new LinkedBlockingQueue<>();
+
+  /** Saves the remove configuration call back functions. */
```
Done

LuQQiu commented on core/server/master/src/main/java/alluxio/master/block/DefaultBlockMaster.java:
```diff
@@ -162,6 +168,17 @@ public Object getFieldValue(MasterWorkerInfo o) {
+  /** Saves the remove configuration call back functions. */
+  private final BlockingQueue<Consumer<Long>> mWorkerLostListeners = new LinkedBlockingQueue<>();
+
+  /** Saves the register new configuration call back functions. */
```
Done

AmplabJenkins commented:
Merged build finished. Test PASSed.

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/ServerConfigurationChecker.java:
```diff
@@ -13,26 +13,34 @@
+  public synchronized void init() {
     mConfMap = new HashMap<>();
-    mLostConfMap = new HashMap<>();
+    mLostNodes = new HashSet<>();
```
We can make `mConfMap` and `mLostNodes` final if we instead use `mConfMap.clear()` and `mLostNodes.clear()`. This also makes it easier for the garbage collector to collect the old configurations.

aaudiber commented on core/server/master/src/main/java/alluxio/master/meta/ServerConfigurationChecker.java:
```diff
@@ -0,0 +1,88 @@
+   * Initializes the default {@link ServerConfigurationChecker}.
+   */
+  public synchronized void init() {
+    mConfMap = new HashMap<>();
```
Minor: We can make `mConfMap` and `mLostNodes` final if we instead use `mConfMap.clear()` and `mLostNodes.clear()`. This also makes it easier for the garbage collector to collect the old configurations.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19458/
Test PASSed.

gpang commented on core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:
```diff
@@ -278,6 +285,7 @@ protected void startMasters(boolean isLeader) {
    */
   protected void stopMasters() {
     try {
+      mWorkerConfigChecker.reset();
```
Should this be moved to `startMasters`? @aaudiber

LuQQiu replied:
By @aaudiber :Actually I agree with Lu that it's better to call init during start instead of during stop. It's better to initialize closer to when we need it so that there's less time for something to interfere. When we call stop() there could still be outstanding RPC threads trying to register workers and modifying the server config checker

@gpang I prefer to change the reset() to startMasters() part to make sure the leader master have empty conf map when it get leadership, PTAL, thanks

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/Alluxio-Pull-Request-Builder/19461/
Test PASSed.
